dbms/src/Disks/DiskS3.cpp

```diff
@@ -0,0 +1,337 @@
+                Poco::FileOutputStream(metadata_path) << s3_path;
+            }
+        }
+
+    private:
+        bool finalized = false;
+        const String metadata_path;
+        const String s3_path;
+    };
+}
+
+DiskS3::DiskS3(String name_, std::shared_ptr<Aws::S3::S3Client> client_, String bucket_, String s3_root_path_, String metadata_path_)
+    : name(std::move(name_))
+    , client(std::move(client_))
+    , bucket(std::move(bucket_))
+    , s3_root_path(std::move(s3_root_path_))
+    , metadata_path(std::move(metadata_path_))
+{
+}
+
+ReservationPtr DiskS3::reserve(UInt64 bytes)
+{
+    if (!tryReserve(bytes))
+        return {};
+    return std::make_unique<DiskS3Reservation>(std::static_pointer_cast<DiskS3>(shared_from_this()), bytes);
+}
+
+bool DiskS3::exists(const String & path) const
+{
+    return Poco::File(metadata_path + path).exists();
+}
+
+bool DiskS3::isFile(const String & path) const
+{
+    return Poco::File(metadata_path + path).isFile();
+}
+
+bool DiskS3::isDirectory(const String & path) const
+{
+    return Poco::File(metadata_path + path).isDirectory();
+}
+
+size_t DiskS3::getFileSize(const String & path) const
+{
+    Aws::S3::Model::GetObjectRequest request;
+    request.SetBucket(bucket);
+    request.SetKey(getS3Path(path));
+    auto outcome = client->GetObject(request);
+    if (!outcome.IsSuccess())
+    {
+        auto & err = outcome.GetError();
+        throw Exception(err.GetMessage(), static_cast<int>(err.GetErrorType()));
+    }
+    else
+    {
+        return outcome.GetResult().GetContentLength();
+    }
+}
+
+void DiskS3::createDirectory(const String & path)
+{
+    Poco::File(metadata_path + path).createDirectory();
+}
+
+void DiskS3::createDirectories(const String & path)
+{
+    Poco::File(metadata_path + path).createDirectories();
+}
+
+DiskDirectoryIteratorPtr DiskS3::iterateDirectory(const String & path)
+{
+    return std::make_unique<DiskS3DirectoryIterator>(metadata_path + path, path);
+}
+
+void DiskS3::moveFile(const String & from_path, const String & to_path)
+{
+    if (exists(to_path))
+        throw Exception("File already exists " + to_path, ErrorCodes::FILE_ALREADY_EXISTS);
+    Poco::File(metadata_path + from_path).renameTo(metadata_path + to_path);
+}
+
+void DiskS3::replaceFile(const String & from_path, const String & to_path)
+{
+    Poco::File from_file(metadata_path + from_path);
+    Poco::File to_file(metadata_path + to_path);
+    if (to_file.exists())
+    {
+        Poco::File tmp_file(metadata_path + to_path + ".old");
+        to_file.renameTo(tmp_file.path());
+        from_file.renameTo(metadata_path + to_path);
+        remove(to_path + ".old");
+    }
+    else
+        from_file.renameTo(to_file.path());
+}
+
+void DiskS3::copyFile(const String & from_path, const String & to_path)
+{
+    if (exists(to_path))
+        remove(to_path);
+
+    String s3_from_path;
+    String s3_to_path = s3_root_path + to_path + getRandomSuffix();
```

Comment:
To better decouple data and metadata I'd generate `s3_to_path` something like  `s3_root_path + '.storage/' + getRandomFilename()`.

Reply:
I had preferred to add origin filename to make bucket content more human-readable and reduce the probability of name overlap.